[jit] serialize modules as classes

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23098 [jit] serialize modules as classes**

This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
[wip][jit] serialize modules as classes

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

zdevito commented on test/jit_utils.py:
```diff
@@ -90,13 +90,16 @@ def _isHookExceptionOk(self, e):
                 if len(m.code) == 0:
-                    # short-circuit if this is an empty module
                     return
+                if isinstance(m, torch._C.ScriptModule):
```
Looks like some tests are just disabled, I am going to only review the non-test parts of this.

zdevito commented on test/test_jit.py:
```diff
@@ -15290,7 +15290,7 @@ def foo(self, x1, x2):
 
             @torch.jit.script_method
-            def wait_script(self, x1, x2):
+            def forward(self, x1, x2):
```
why did this change?

suo replied:
EmitModuleHook now compares the `.code` of the modules instead of relying on internal details of the namespace. But `.code` gives you the `forward()` method, otherwise it doesn't work.

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -696,11 +690,15 @@ void ScriptModuleSerializer::convertModel(
+  pickler.start();
+  pickler.addIValue(module.module_object());
+  pickler.finish();
+  writer_.writeRecord("module.pkl", pickler.stack().data(), pickler.stack().size());
```
Let's not use the work "module". Afterall this archieve doesn't need to encode modules. How about `code/` for the code, and `data.pkl` for the data?

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -194,25 +194,111 @@ struct SourceImporter {
-          definitions.emplace_back(Def(method_def));
-          resolvers.emplace_back(resolver_);
+        std::vector<Assign> attributes;
+
```
This function is really long. It has two phases, parsing the statement body and then acting on that parse. Can we split this up?

suo replied:
going to follow up on this because there are a bunch of downstream changes

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -122,6 +122,7 @@ const static std::unordered_set<std::string> reserved_names = {
     "aten",
     "attribute",
     "CONSTANTS",
+    "PICKLES",
```
Where is this used?

suo replied:
this was left in from a previous change, will remove

zdevito reviewed (approved):
Cool! Looks good. Just minor comments. Obviously needs back-compat support as discussed in person.

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"

[wip][jit] serialize modules as classes

gh-metadata: pytorch pytorch 23098 gh/suo/99/head

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head

Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[wip][jit] serialize modules as classes"


gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)

suo committed:
Update on "[jit] serialize modules as classes"


This PR introduces a new way to serialize models. Before, the module
hierarchy was described using a json file. Now, all modules are
described by code, and pickled as an object. So the process of
deserializing a model is just unpickling the object.

The new zip archive should look like:
```
model_file  # zip file
|-- model.json  # contains tensor table, soon to be removed
|-- code/
    |-- ..various .py files describing modules and classes
|-- data.pkl  # pickle archive of module data
|-- tensors/
    |-- tensor data
```

gh-metadata: pytorch pytorch 23098 gh/suo/99/head


Differential Revision: [D16383328](https://our.internmc.facebook.com/intern/diff/D16383328)
